doctr/models/recognition/parseq/pytorch.py

```diff
@@ -0,0 +1,503 @@
+        null_ctx = self.embed(target[:, :1])
+        content = self.pos_queries[:, : sequence_length - 1] + self.embed(target[:, 1:])
+        content = self.dropout(torch.cat([null_ctx, content], dim=1))
+        if target_query is None:
+            target_query = self.pos_queries[:, :sequence_length].expand(batch_size, -1, -1)
+        target_query = self.dropout(target_query)
+        return self.decoder(target_query, content, memory, target_mask)
+
+    def decode_autoregressive(self, features: torch.Tensor) -> torch.Tensor:
+        """Generate predictions for the given features."""
+        # Padding symbol + SOS at the beginning
+        ys = torch.full(
+            (features.size(0), self.max_length), self.vocab_size + 2, dtype=torch.long, device=features.device
+        )  # pad
+        ys[:, 0] = self.vocab_size + 1  # SOS token
+        pos_queries = self.pos_queries[:, : self.max_length + 1].expand(features.size(0), -1, -1)
+        # Create query mask for the decoder attention
+        query_mask = (
+            torch.tril(torch.ones((self.max_length + 1, self.max_length + 1), device=features.device), diagonal=0).to(
+                dtype=torch.bool
+            )
+        ).int()
+
+        pos_logits = []
+        for i in range(self.max_length):
+            # Decode one token at a time without providing information about the future tokens
+            tgt_out = self.decode(
+                ys[:, : i + 1],
+                features,
+                query_mask[i : i + 1, : i + 1],
+                target_query=pos_queries[:, i : i + 1],
+            )
+            pos_prob = self.head(tgt_out)
+            pos_logits.append(pos_prob)
+
+            if i + 1 < self.max_length:
+                # Update with the next token
+                ys[:, i + 1] = pos_prob.squeeze().argmax(-1)
+
+                # Stop decoding if all sequences have reached the EOS token
+                if (ys == self.vocab_size).any(dim=-1).all():
+                    break
+
+        logits = torch.cat(pos_logits, dim=1)  # (N, max_length, vocab_size + 1)
+
+        # One refine iteration
+        # Update query mask
+        query_mask[
+            torch.triu(
+                torch.ones(self.max_length + 1, self.max_length + 1, dtype=torch.bool, device=features.device), 2
+            )
+        ] = 1
+
+        # Prepare target input for 1 refine iteration
+        sos = torch.full((features.size(0), 1), self.vocab_size + 1, dtype=torch.long, device=features.device)
+        ys = torch.cat([sos, logits[:, :-1].argmax(-1)], dim=1)
+
+        # Create padding mask for refined target input maskes all behind EOS token as False
+        # (N, 1, 1, max_length)
+        target_pad_mask = ~((ys == self.vocab_size).int().cumsum(-1) > 0).unsqueeze(1).unsqueeze(1)
+        mask = (target_pad_mask.bool() & query_mask[:, : ys.shape[1]].bool()).int()
+        logits = self.head(self.decode(ys, features, mask, target_query=pos_queries))
```

Comment:
The refinement process can be done regardless of the initial decoding scheme (AR or NAR). I suggest moving this to a separate method so it can be used by either AR or NAR decoding.

Reply:
👍